src/util.rs

```diff
@@ -5,7 +5,280 @@
+    /// TODO: What about reads/mutation via raw pointers? Presumably these can
+    /// happen under the following conditions:
+    /// - Mutation only occurs inside `UnsafeCell`s
+    /// - Reads only happen using `UnsafeCell`s in places where there are
+    ///   `UnsafeCell`s in `T` (otherwise, those reads could be unsound due to
+    ///   assuming no concurrent mutation)
+    ///
+    /// [`UnsafeCell`]: core::cell::UnsafeCell
+    pub(crate) unsafe fn _as_ref(&self) -> &'a T {
+        // TODO(#429): Add a safety comment. This will depend on how we resolve
+        // the question about how to define the safety invariants on this
+        // method.
+        //
+        // Old draft of safety comment:
+        // - By invariant, `self.ptr` is properly-aligned for `T`.
+        // - By invariant, `self.ptr` is "dereferenceable" in that it points to
+        //   a single allocation
+        // - By invariant, the allocation is live for `'a`
+        // - The caller promises that no mutable references exist to this region
+        //   during `'a`
+        // - The caller promises that `UnsafeCell`s match exactly
+        // - The caller promises that the memory region contains a
+        //   validly-intialized `T`
+        #[allow(clippy::undocumented_unsafe_blocks)]
+        unsafe {
+            self.ptr.as_ref()
+        }
+    }
+}
+
+impl<'a, T: 'a> Ptr<'a, [T]> {
+    pub(crate) fn _len(&self) -> usize {
+        // TODO(#67): Remove this allow. See NonNullSliceExt for more details.
+        #[allow(unstable_name_collisions)]
+        self.ptr.len()
+    }
+}
+
+impl<'a> Ptr<'a, [u8]> {
+    /// Attempts to cast `self` to a `U` using the given cast type.
+    ///
+    /// Returns `None` if the resulting `U` would be invalidly-aligned or if no
+    /// `U` can fit in `self`. On success, returns a pointer to the
+    /// largest-possible `U` which fits in `self`.
+    ///
+    /// # Safety
+    ///
+    /// The caller may assume that this implementation is correct, and may rely
+    /// on that assumption for the soundness of their code. In particular, the
+    /// caller may assume that, if `try_cast_into` returns `Some((ptr,
+    /// split_at))`, then:
+    /// - If this is a prefix cast, `ptr` refers to the byte range `[0,
+    ///   split_at)` in `self`.
+    /// - If this is a suffix cast, `ptr` refers to the byte range `[split_at,
+    ///   self.len())` in `self`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `U` is a DST whose trailing slice element is zero-sized.
+    pub(crate) fn _try_cast_into<U: 'a + ?Sized + KnownLayout>(
+        &self,
+        cast_type: _CastType,
+    ) -> Option<(Ptr<'a, U>, usize)> {
+        // PANICS: By invariant, the byte range addressed by `self.ptr` does not
+        // wrap around the address space. This implies that the sum of the
+        // address (represented as a `usize`) and length do not overflow
+        // `usize`, as required by `validate_cast_and_convert_metadata`.
```

Comment:
Considering that "SAFETY" is used to introduce a comment that explains why code is actually safe, it's a bit odd that "PANICS" is introducing a comment about why something *doesn't* panic. I think you could just append "and thus `validate_cast_and_convert_metadata` won't panic.

Reply:
Done.